temporal-sdk/src/main/java/io/temporal/internal/worker/WorkflowRunLockManager.java

```diff
@@ -19,89 +19,101 @@
 
 package io.temporal.internal.worker;
 
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Date;
 import java.util.HashMap;
-import java.util.concurrent.locks.Lock;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
 import java.util.concurrent.locks.ReentrantLock;
 
 final class WorkflowRunLockManager {
-
-  private static class CountableLock {
-    private final Lock lock = new ReentrantLock();
-    private int count = 1;
-
-    void incrementCount() {
-      count++;
-    }
-
-    void decrementCount() {
-      count--;
-    }
-
-    int getCount() {
-      return count;
-    }
-
-    Lock getLock() {
-      return lock;
-    }
-  }
-
-  private final Lock mapLock = new ReentrantLock();
-  private final HashMap<String, CountableLock> perRunLock = new HashMap<>();
+  // This is a single lock for the whole worker.
+  // It's ok, because it's acquired for a very short time.
+  // If this ever becomes a bottleneck, consider:
+  //   - rework this lock to use com.google.common.util.concurrent.Striped and
+  //   - make `perRunLock` a ConcurrentHashMap to increase parallelism of this code
+  private final ReentrantLock mapLock = new ReentrantLock();
+  private final Map<String, LockData> perRunLock = new HashMap<>();
 
   /**
-   * This method returns a lock that can be used to serialize workflow task processing for a
-   * particular workflow run. This is used to make sure that query tasks and real workflow tasks are
-   * serialized when sticky is on.
-   *
-   * @param runId
-   * @return a lock to be used during workflow task processing
+   * @param runId to take a lock for
+   * @param time the maximum time to wait for the lock
+   * @param unit the time unit of the time argument
+   * @return true if the lock is taken
    */
-  Lock getLockForLocking(String runId) {
+  boolean tryLock(String runId, long time, TimeUnit unit) throws InterruptedException {
     mapLock.lock();
-
     try {
-      CountableLock cl = perRunLock.get(runId);
-      if (cl == null) {
-        cl = new CountableLock();
-        perRunLock.put(runId, cl);
-      } else {
-        cl.incrementCount();
+      Date deadline = new Date(System.currentTimeMillis() + unit.toMillis(time));
+      boolean stillWaiting = true;
+      LockData lockData = tryPersistLockByCurrentThreadLocked(runId);
+      while (Thread.currentThread() != lockData.thread) {
```

Comment:
From what I'm reading, this will never evaluate to true on successive iterations. How can the map remove an old lock data (i.e. invoking/persisting `computeIfAbsent` result) if you've locked the entire map here with `mapLock`? If this is just for the first invocation, an `if` is clearer.

Reply:
Spurious wake-ups